File: FloodFill(DFS).java
// Time Complexity :O(m*n)
// Space Complexity :O(m*n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


class Solution {
    int [][] dirs;
    int oldColor;
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if( image == null || image.length == 0 || image[sr][sc] == newColor)return image;
        oldColor = image[sr][sc];
        dirs = new int [][] {{1,0},{0,1},{-1,0},{0,-1}};
        dfs(image,sr,sc,newColor);
        return image;       
    }
    private void dfs(int[][] image, int sr, int sc, int newColor){
        
        if(sr>=image.length || sc >= image[0].length || sr < 0 || sc < 0 || image[sr][sc] != oldColor)return;
        
        image[sr][sc] = newColor;
        for(int [] dir : dirs){
            int r = dir[0] + sr;
            int c = dir[1] + sc;
            dfs(image,r,c,newColor);
        }
    }
}

File: Flood_Fill(DFS).cpp
//Time Complexity-O(m*n) - number of pixels
//Space Complexity-O(m*n)- recursive stack Space
//Did the code run on Leetcode? yes

class Solution {
public:
    vector<vector<int>>directions={{0,1},{1,0},{0,-1},{-1,0}};
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        int tempcolor=image[sr][sc];
        if(newColor==tempcolor)
        {
            return image;
        }
        flood(image,sr,sc,newColor,tempcolor);
        return image;
    }
    void flood(vector<vector<int>>& image, int sr,int sc,int newColor,int temp)
    {
        if(image[sr][sc]!=temp)
        {
            return;
        }
        image[sr][sc]=newColor;
        for(int i=0;i<directions.size();i++)
        {
            vector<int>k=directions[i];
            int row=k[0]+sr;
            int col=k[1]+sc;
            if(row>=0 && row<image.size() && col>=0 &&col<image[0].size())
            {
                  flood(image,row,col,newColor,temp);
            }
        }
    }
};